### PR TITLE
Improve error handling and reporting when deleting items.

### LINKED ISSUE #355: heketi-cli: proper error message for cluster/node/device delete
It is not possible to delete cluster without deleting nodes, devices and volumes.  

=> docker exec heketi heketi-cli --server http://localhost:8080 cluster delete 46d613cf778d1a11ab2f36a2dd3dc803
Error: Conflict

Logs:
[negroni] Started DELETE /clusters/46d613cf778d1a11ab2f36a2dd3dc803
[heketi] WARNING 2016/05/27 12:01:01 Unable to delete cluster [46d613cf778d1a11ab2f36a2dd3dc803] because it contains volumes and/or nodes
[negroni] Completed 409 Conflict in 466.033µs

=>docker exec heketi heketi-cli --server http://localhost:8080 node delete 53bb0c58ceaa2042c5088377de9c872f
Error: Conflict

Logs:
[negroni] Started DELETE /nodes/53bb0c58ceaa2042c5088377de9c872f
[negroni] Completed 409 Conflict in 323.013µs

=> docker exec heketi heketi-cli --server http://localhost:8080 device delete 77fddc8c8ed815ab09965095941cf899
Error: Conflict

Logs:
[negroni] Started DELETE /devices/77fddc8c8ed815ab09965095941cf899
[negroni] Completed 409 Conflict in 368.05µs

Should display proper error message to STDOUT.
For nodes and devices it doesn't log proper message in logs too. Creating a new issue to track it because if not wrong https://github.com/heketi/utils/commit/34e5a97154d1d016635ad6ce436a68fdc8a75876 will not fix this one.

### PR DESCRIPTION
**Pull request by jarrpa**
Allow more useful error messages to be reported and logged when attempting to
delete clusters, nodes, and devices in the glusterfs app. Minor changes also
to simplify and bring consistency to error reporting within related functions.

Closes #355

Signed-off-by: Jose A. Rivera <jarrpa@redhat.com>

### CODE REVIEW

File: apps/glusterfs/app_cluster.go [hunk 1]
```diff
@@ -139,16 +141,17 @@ func (a *App) ClusterDelete(w http.ResponseWriter, r *http.Request) {
...
 
 		err = entry.Delete(tx)
-		if err == ErrConflict {
-			http.Error(w, err.Error(), http.StatusConflict)
-			return err
-		} else if err != nil {
-			http.Error(w, err.Error(), http.StatusInternalServerError)
-			return err
+		if err != nil {
+			if strings.Contains(err.Error(), "Unable to delete") {
```
Comment (lpabon) [inline apps/glusterfs/app_cluster.go]: If an error is being checked, then it must be a variable saved in errors.go

File: apps/glusterfs/cluster_entry_test.go [hunk 2]
```diff
@@ -248,7 +249,7 @@ func TestNewClusterEntrySaveDelete(t *testing.T) {
 		return nil
 
 	})
-	tests.Assert(t, err == ErrConflict)
+	tests.Assert(t, strings.Contains(err.Error(), "Unable to delete"))
```
Comment (lpabon) [inline apps/glusterfs/cluster_entry_test.go]: Same here

File: apps/glusterfs/cluster_entry_test.go [hunk 3]
```diff
@@ -277,7 +278,7 @@ func TestNewClusterEntrySaveDelete(t *testing.T) {
...
```
Comment (lpabon) [inline apps/glusterfs/cluster_entry_test.go]: Same here

**Comment by lpabon** [comment]
The change is based on the style of checking for strings in err.  That should be avoided.  If an error is going to be checked, then it must have a value.  Most of these are stored in errors.go.

There are parts of the fix which add logging to errors, those are cool, but changing ErrConflict does not seem to add benefit.

**Comment by jarrpa** [comment]
But the entire point of the issue is that the error message is useless. It is also impossible to check the current useful error message against a constant because the error message is generated per ID. The main alternative would be to remove ErrConflict entirely and create unique conflict errors for each entity that needs them. We would still lose the ID of the conflicting item, but that might be an okay compromise?

**Comment by lpabon** [comment]
Let's see if there is another way to fix this issue.  Let's talk on Monday.

File: apps/glusterfs/app_cluster.go [hunk 4]
```diff
@@ -143,12 +143,13 @@ func (a *App) ClusterDelete(w http.ResponseWriter, r *http.Request) {
...
-		} else if err != nil {
-			http.Error(w, err.Error(), http.StatusInternalServerError)
-			return err
+		if err != nil {
+			if err == ErrConflict {
+				http.Error(w, entry.ErrConflictFmt(), http.StatusConflict)
+			} else {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+			}
+			return logger.Err(err)
```
Comment (lpabon) [inline apps/glusterfs/app_cluster.go]: This is cool, but like the #503 we should not log ERROR on conflict.  ERROR is more of a "Hey this is messed up, fix this, i don't know what to do!".  Also, although I didn't point this out in #503, most likely there is no need to log errors in app_XXX .

File: apps/glusterfs/node_entry.go [hunk 5]
```diff
@@ -159,13 +159,16 @@ func (n *NodeEntry) IsDeleteOk() bool {
 	return true
 }
 
+func (n *NodeEntry) ErrConflictFmt() error {
```
Comment (lpabon) [inline apps/glusterfs/node_entry.go]: Interesting style, but I would name it something different because ErrXXX is normally for errors and XXXFmt() is normally for something like the string in Printf( "fmt..").  How about:

```go
func (n *XXX) ErrorMsg( err error ) string {

   switch err {
   case ErrConflict:
       return fmt.Sprintf("...")
...
   }
}
```

Comment (lpabon) [inline apps/glusterfs/node_entry.go]: Or you can call it ErrorInfo, ErrorString... you get the idea.

File: apps/glusterfs/node_entry.go [hunk 6]
```diff
@@ -159,13 +159,16 @@ func (n *NodeEntry) IsDeleteOk() bool {
...
+}
+
 func (n *NodeEntry) Delete(tx *bolt.Tx) error {
 	godbc.Require(tx != nil)
 
 	// Check if the nodes still has drives
 	if !n.IsDeleteOk() {
-		logger.Warning("Unable to delete node [%v] because it contains devices", n.Info.Id)
-		return ErrConflict
+		return logger.Err(ErrConflict)
```
Comment (lpabon) [inline apps/glusterfs/node_entry.go]: Here, this is not an ERROR.  Same as the others.  You may want to have logger.Warning (and others) return `error`.  You can do that in another PR, and then have this one use that.

File: pkg/utils/log.go [hunk 7]
```diff
@@ -138,6 +138,15 @@ func (l *Logger) Warning(format string, v ...interface{}) {
 	}
 }
 
+// Log error variable as a warning
+func (l *Logger) Warn(err error) error {
```
Comment (lpabon) [inline pkg/utils/log.go]: Awesome.  This needs unit tests.  You may want to name the ones that take error something similar.  For example, we already have `Err()` and `LogError()`, so this could be `WarnErr()` or `WarningError()` or something like that.

File: apps/glusterfs/node_entry.go [hunk 8]
```diff
@@ -164,8 +164,7 @@ func (n *NodeEntry) Delete(tx *bolt.Tx) error {
 
 	// Check if the nodes still has drives
 	if !n.IsDeleteOk() {
-		logger.Warning("Unable to delete node [%v] because it contains devices", n.Info.Id)
```
Comment (lpabon) [inline apps/glusterfs/node_entry.go]: We are loosing logging information here.  These lines should not be removed.

File: apps/glusterfs/device_entry.go [hunk 9]
```diff
@@ -153,8 +153,7 @@ func (d *DeviceEntry) Delete(tx *bolt.Tx) error {
 
 	// Check if the devices still has drives
 	if !d.IsDeleteOk() {
-		logger.Warning("Unable to delete device [%v] because it contains bricks", d.Info.Id)
```
Comment (lpabon) [inline apps/glusterfs/device_entry.go]: We are loosing logging information here.  These lines should not be removed.

File: apps/glusterfs/cluster_entry.go [hunk 10]
```diff
@@ -80,10 +80,9 @@ func (c *ClusterEntry) Save(tx *bolt.Tx) error {
 func (c *ClusterEntry) Delete(tx *bolt.Tx) error {
 	godbc.Require(tx != nil)
 
-	// Check if the nodes still has drives
+	// Check if the cluster still has nodes or volumes
 	if len(c.Info.Nodes) > 0 || len(c.Info.Volumes) > 0 {
-		logger.Warning("Unable to delete cluster [%v] because it contains volumes and/or nodes", c.Info.Id)
```
Comment (lpabon) [inline apps/glusterfs/cluster_entry.go]: We are loosing logging information here.  These lines should not be removed.

File: apps/glusterfs/app_node.go [hunk 11]
```diff
@@ -227,7 +228,7 @@ func (a *App) NodeDelete(w http.ResponseWriter, r *http.Request) {
 
 		// Check the node can be deleted
 		if !node.IsDeleteOk() {
-			http.Error(w, ErrConflict.Error(), http.StatusConflict)
+			http.Error(w, fmt.Sprintf("Unable to delete node [%v] because it contains devices", node.Info.Id), http.StatusConflict)
```
Comment (lpabon) [inline apps/glusterfs/app_node.go]: Good, this needs unit test to check.

File: apps/glusterfs/app_device.go [hunk 12]
```diff
@@ -226,7 +227,7 @@ func (a *App) DeviceDelete(w http.ResponseWriter, r *http.Request) {
 
 		// Check if we can delete the device
 		if !device.IsDeleteOk() {
-			http.Error(w, ErrConflict.Error(), http.StatusConflict)
+			http.Error(w, fmt.Sprintf("Unable to delete device [%v] because it contains bricks", device.Info.Id), http.StatusConflict)
```
Comment (lpabon) [inline apps/glusterfs/app_device.go]: Good, this needs unit test to check.

File: apps/glusterfs/app_cluster.go [hunk 13]
```diff
@@ -143,11 +145,12 @@ func (a *App) ClusterDelete(w http.ResponseWriter, r *http.Request) {
...
 
 		err = entry.Delete(tx)
-		if err == ErrConflict {
-			http.Error(w, err.Error(), http.StatusConflict)
-			return err
-		} else if err != nil {
-			http.Error(w, err.Error(), http.StatusInternalServerError)
+		if err != nil {
+			if err == ErrConflict {
+				http.Error(w, fmt.Sprintf("Unable to delete cluster [%v] because it contains volumes and/or nodes", entry.Info.Id), http.StatusConflict)
```
Comment (lpabon) [inline apps/glusterfs/app_cluster.go]: Good, this needs unit test to check.

File: pkg/utils/log.go (hunk 7 above)
Comment (lpabon) [inline pkg/utils/log.go]: But you may want to do this in another PR.  I do not think you need this in this PR.

**Comment by jarrpa** [comment]
Moved the first patch into #509.

**Comment by jarrpa** [comment]
Regarding the loss of logging information: would a reasonable compromise be to change the error text of ErrConflict from "Conflict" to "The target either contains other items or is in use."? Or do you want to maintain the current level of verbosity in the logs for the XXX_entry.go files?

File: apps/glusterfs/app_cluster.go [hunk 14]
```diff
@@ -18,10 +18,12 @@ package glusterfs
 
 import (
 	"encoding/json"
+	"fmt"
+	"net/http"
+
```
Comment (lpabon) [inline apps/glusterfs/app_cluster.go]: Nice! thanks :smile:

**Comment by lpabon** [comment]
I like both :smile: .  Make the change to change from just `Conflict` to `The target either contains other items or is in use`.  And keep the current logs because they log the ID which had the issue.

**Comment by lpabon** [comment]
One more thing.  These changes need unit tests updates to verify them.

**Comment by jarrpa** [comment]
Did both, added tests. I don't directly test the ConflictString() functions right now, since they are one-line functions and the test would just be to copy out the one line. Instead I test they work by calling them for the verification of the REST response body content.

File: apps/glusterfs/app_cluster_test.go [hunk 15]
```diff
@@ -334,20 +334,23 @@ func TestClusterDelete(t *testing.T) {
...
 	tests.Assert(t, r.StatusCode == http.StatusConflict)
+	tests.Assert(t, utils.GetErrorFromResponse(r).Error() == entries[1].ConflictString())
 
 	// Check that we cannot delete a cluster with nodes
 	req, err = http.NewRequest("DELETE", ts.URL+"/clusters/"+"a3", nil)
 	tests.Assert(t, err == nil)
 	r, err = http.DefaultClient.Do(req)
 	tests.Assert(t, err == nil)
 	tests.Assert(t, r.StatusCode == http.StatusConflict)
+	tests.Assert(t, utils.GetErrorFromResponse(r).Error() == entries[2].ConflictString())
```
Comment (lpabon) [inline apps/glusterfs/app_cluster_test.go]: nice

File: apps/glusterfs/cluster_entry.go [hunk 16]
```diff
@@ -77,12 +78,16 @@ func (c *ClusterEntry) Save(tx *bolt.Tx) error {
...
+}
+
 func (c *ClusterEntry) Delete(tx *bolt.Tx) error {
 	godbc.Require(tx != nil)
 
-	// Check if the nodes still has drives
+	// Check if the cluster still has nodes or volumes
 	if len(c.Info.Nodes) > 0 || len(c.Info.Volumes) > 0 {
-		logger.Warning("Unable to delete cluster [%v] because it contains volumes and/or nodes", c.Info.Id)
+		logger.Warning(c.ConflictString())
```
Comment (lpabon) [inline apps/glusterfs/cluster_entry.go]: Ah nice!

**Comment by lpabon** [comment]
Code looks good.  I just need to verify it.  I'll get to that tomorrow.

**Comment by jarrpa** [comment]
Great! Once this gets merged we should probably also close #356.

**Comment by lpabon** [comment]
Awesome!

```
$ ./heketi-cli -s http://localhost:8080 node delete c2304ee9667c36ad4dc6b0c11b954bb6
Error: Unable to delete node [c2304ee9667c36ad4dc6b0c11b954bb6] because it contains devices

$ ./heketi-cli -s http://localhost:8080 cluster delete 9d525bf98fa9c20816c3cdcfdfa70f64
Error: Unable to delete cluster [9d525bf98fa9c20816c3cdcfdfa70f64] because it contains volumes and/or nodes
```

**Comment by lpabon** [comment]
Closes #356
